### PR TITLE
Notification of available haxelib version

### PR DESCRIPTION
**Pull request by ibilon**
Implementation of #282 

I added the check to all command that connects to the internet already (with the exception of install which don't always do).

### CODE REVIEW

**Comment by ibilon** [comment]
Ok so maybe not a safe assumption that "haxelib_client" always exists :)

**Comment by Simn** [comment]
On an unrelated random note, either you or Mark have to change your GitHub avatar or I'll start talking to the wrong person soon...

File: src/haxelib/SemVer.hx [hunk 1]
```diff
@@ -97,6 +97,12 @@ abstract SemVer(String) to String {
 	inline function get_valid()
 		return isValid(this);
 
+	@:op(a > b) static function greaterThan(a:SemVer, b:SemVer) {
```
Comment (nadako) [inline src/haxelib/SemVer.hx]: I think this is wrong, since it doesn't check `preview`/`previewNum` fields. There's `Semver.compare(a,b)` that will return 1 if a > b.

File: src/haxelib/client/Main.hx [hunk 2]
```diff
@@ -154,27 +154,27 @@ class Main {
 		args = Sys.args();
 
 		commands = new List();
-		addCommand("install", install, "install a given library, or all libraries from a hxml file", Basic);
-		addCommand("update", update, "update a single library (if given) or all installed libraries", Basic);
+		addCommand("install", checkUpdateThen.bind(install), "install a given library, or all libraries from a hxml file", Basic);
```
Comment (nadako) [inline src/haxelib/client/Main.hx]: I think this should be done differently. We have `net` property for commands, so the check should be simply added after `loadProxy` here: https://github.com/HaxeFoundation/haxelib/blob/development/src/haxelib/client/Main.hx#L366

**Comment by ibilon** [comment]
Thanks for the feedback @nadako 
I'll have to look at the source code, seems I'm missing quite some knowledge about haxelib.

**Comment by nadako** [comment]
Np. I wonder though if it's better to add `SiteApi.getLatestVersion` that just returns a version string instead of using `infos` that returns the whole project information. Concerning both traffic and server db load.

**Comment by ibilon** [comment]
I thought about that but I was unsure on the effect of modifying SiteApi, even when just adding something.

**Comment by nadako** [comment]
It shouldn't break anything, I think, but like any API, it should be carefully designed. To me, it feels a bit weird to have `getLatestVersion` for what's already possible to do with the `infos` method, but OTOH it has a specific goal: to return latest library version in the most efficient way, so maybe it's okay.

I'll just ping @Simn, @ncannasse and @back2dos and hear what they think. :)

**Comment by ibilon** [comment]
I checked just to be sure, and yes adding a function doesn't break an old client.

**Comment by nadako** [comment]
Right, update also uses `infos().getLatest()` which can be replaced by the new shiny method. I'd say let's do it!

**Comment by ibilon** [comment]
Did a try on it, had to go sql for that since I couldn't figure out how to order on several column otherwise.

File: src/haxelib/client/Main.hx [hunk 3]
```diff
@@ -1037,17 +1048,17 @@ class Main {
 		}
 		else
 		{
-			var inf = try site.infos(p) catch( e : Dynamic ) { Sys.println(e); return; };
-			p = inf.name;
```
Comment (nadako) [inline src/haxelib/client/Main.hx]: This one was needed, actually, see f358bfada2812970f94986defe6f2108d232dfe2, gotta look into how to get that name in proper case without `infos` call.

File: src/haxelib/server/Repo.hx [hunk 4]
```diff
@@ -80,6 +80,15 @@ class Repo implements SiteApi {
 		};
 	}
 
+	public function getLatestVersion( project : String ) : SemVer {
+		var p = Project.manager.select($name == project);
+		if( p == null )
+			throw "No such Project : "+project;
+
+		var v = Version.manager.unsafeObjects('SELECT * FROM Version WHERE project == ${p.id} ORDER BY major DESC, minor DESC, patch DESC, previewNum DESC LIMIT 1', false);
```
Comment (nadako) [inline src/haxelib/server/Repo.hx]: I think this is not correct, since it doesn't account `preview` field that should be also considered for sorting. Not sure if it can be done within a simple SQL query.

File: src/haxelib/server/Repo.hx [hunk 5]
```diff
@@ -80,6 +80,15 @@ class Repo implements SiteApi {
 		};
 	}
 
+	public function getLatestVersion( project : String ) : SemVer {
+		var p = Project.manager.select($name == project);
+		if( p == null )
+			throw "No such Project : "+project;
+
+		var vl = Version.manager.search($project == p.id, { orderBy: [-major, -minor, -patch, -previewNum], limit: 1});
```
Comment (nadako) [inline src/haxelib/server/Repo.hx]: Same comment as for the previous version:
> I think this is not correct, since it doesn't account preview field that should be also considered for sorting. Not sure if it can be done within a simple SQL query.

Comment (ibilon) [inline src/haxelib/server/Repo.hx]: Oh right, for some reason I thought those where the same field.
Will have to check some sql, no idea on how enums work there.

File: src/haxelib/client/Main.hx (hunk 3 above)
Comment (ibilon) [inline src/haxelib/client/Main.hx]: Oh so there was a reason, thought it was weird to do that if the value was identical.
It a problem though :/ and adding a `realName` function to SiteApi just for that seems suboptimal, would have the merit to be explicit at least. Wouldn't happen to be any other places that do that?

Comment (ibilon) [inline src/haxelib/client/Main.hx]:
```
var latest = try site.getLatestVersion(p) catch( e : Dynamic ) { Sys.println(e); return; };
p = p.toLowerCase();
p = FileSystem.readDirectory(pdir).filter(function (dir) return dir.toLowerCase() == p)[0];
```
that would be too ugly I guess?

If only the folder was made in lower case to begin with :/

Comment (nadako) [inline src/haxelib/client/Main.hx]: You probably meant `rep` instead of `pdir`, but yeah, that might work! Please factor that code into a separate function tho :)

File: src/haxelib/server/Repo.hx (hunk 5 above)
Comment (nadako) [inline src/haxelib/server/Repo.hx]: Gotta check SPOD macros, maybe enum indices are used, then you can sort on them easily.

**Comment by ibilon** [comment]
Updated getLatestVersion, had to go back to sql.

They are indeed stored as int,
so you can easily sort them, or you should but they are nullable and null is below all int,
but 3.2.0 > 3.2.0-rc.

Sql has a function `isnull(var, val)` which replace var with val if it's null,
can use the same values as SemVer.compare => 100 for preview and -1 for previewNumber (so not needed).

But sqlite named it `ifnull` otherwise that'd be too easy :)

**Comment by nadako** [comment]
Fails OSX now: https://travis-ci.org/HaxeFoundation/haxelib/jobs/114881349#L3845

**Comment by ibilon** [comment]
Fixed it, apparently isnull is something, works with ifnull on both mysql and sqlite (that's even better actually).

**Comment by andyli** [comment]
I agree it would be useful, but please don't merge it until someone has high confidence that selfupdate actually works for all configurations.
If haxelib keeps telling ppl to do selfupdate when it actually doesn't work it would be funny... ; )

**Comment by nadako** [comment]
Right, let's deal with selfupdate in #172

**Comment by ibilon** [comment]
With #284 should the message be changed to `Do "haxelib update haxelib_client to get the latest version`?

(update installs the latest version even if the lib isn't installed)

Since that's basically all that `selfupdate` would do.

Personally I'm not fan of redundant commands like upgrade/update, install/local/

**Comment by nadako** [comment]
One more issue to think about: currently we duplicate version number in `haxelib.json` and in `Main.VERSION` which is prone to mistakes. E.g. you fix stuff, change `haxelib.json`, upload a new version, but forgot to update the `VERSION` field, and you'll be getting a message about updating to the current version, which is just plain retarded).

I see two immediate solutions for this:
 * Load the `VERSION` value from `haxelib.json` using macro.

or vise-versa...

 * Build `haxelib.json` from a template in our `Package.hx` script, reading version from `Main.VERSION`.

I think second one is the most error-proof, since we eliminate the case where we first build the `run.n` with the old `haxelib.json` version and then change `haxelib.json` :)

File: src/haxelib/SemVer.hx [hunk 6]
```diff
@@ -97,6 +97,10 @@ abstract SemVer(String) to String {
 	inline function get_valid()
 		return isValid(this);
 
+	@:op(a > b) static inline function greaterThan(a:SemVer, b:SemVer) {
```
Comment (nadako) [inline src/haxelib/SemVer.hx]: It feels a bit weird to only have `>` operator and not the other ones. For a single case, maybe it's simplier to just use `compare` method directly instead of `>`, even though it doesn't look great... I'm not sure though.

File: src/haxelib/client/Main.hx [hunk 7]
```diff
@@ -188,6 +188,14 @@ class Main {
 	}
 
 
+	function checkUpdate() {
+		var latest:SemVer = try site.getLatestVersion("haxelib_client") catch (_:Dynamic) null;
+
+		if (latest != null && latest > VERSION) {
+			print('A new version (${latest}) of haxelib is available.\nDo "haxelib selfupdate" to get the latest version.\n');
```
Comment (nadako) [inline src/haxelib/client/Main.hx]: Don't forget to reword this to `haxelib update haxelib` as we want to deprecate `selfupdate`

File: src/haxelib/client/Main.hx [hunk 8]
```diff
@@ -188,6 +188,14 @@ class Main {
 	}
 
 
+	function checkUpdate() {
+		var latest:SemVer = try site.getLatestVersion("haxelib_client") catch (_:Dynamic) null;
```
Comment (nadako) [inline src/haxelib/client/Main.hx]: The library is going to be `haxelib` instead of `haxelib_client` as were discussed.

**Comment by nadako** [comment]
I added a couple comments here. Could you please rebase and squash this?

File: src/haxelib/client/Main.hx [hunk 9]
```diff
@@ -173,8 +173,8 @@ class Main {
 		addCommand("local", local, "install the specified package locally", Development, false);
 		addCommand("dev", dev, "set the development directory for a given library", Development, false);
 		//TODO: generate command about VCS by Vcs.getAll()
-		addCommand("git", function()doVcs(VcsID.Git), "use Git repository as library", Development);
-		addCommand("hg", function()doVcs(VcsID.Hg), "use Mercurial (hg) repository as library", Development);
+		addCommand("git", function() doVcs(VcsID.Git), "use Git repository as library", Development);
+		addCommand("hg", function() doVcs(VcsID.Hg), "use Mercurial (hg) repository as library", Development);
```
Comment (nadako) [inline src/haxelib/client/Main.hx]: unrelated whitespace change, but oh well...

File: src/haxelib/SiteApi.hx [hunk 10]
```diff
@@ -26,6 +26,7 @@ import haxelib.Data;
 interface SiteApi {
 	public function search( word : String ) : List<{ id : Int, name : String }>;
 	public function infos( project : String ) : ProjectInfos;
+	public function getLatestVersion( project : String) : SemVer;
```
Comment (nadako) [inline src/haxelib/SiteApi.hx]: looks like a space is missing before the `)` here :)

File: src/haxelib/client/Main.hx [hunk 11]
```diff
@@ -1037,19 +1049,32 @@ class Main {
...
+				setCurrent(p, latest, true);
 		}
 	}
+
+	function projectNameToDir( project:String, state ) {
+		var p = project.toLowerCase();
+
+		var l = FileSystem.readDirectory(state.rep).filter(function (dir) return dir.toLowerCase() == p);
+
+		switch (l.length) {
```
Comment (nadako) [inline src/haxelib/client/Main.hx]: it's not important or anything, but note that this could be written as
```haxe
switch (l) {
    case []: throw ...;
    case [dir]: return Data.unsafe(dir);
    case _: throw ...;
}
```
the generated code will be pretty much the same, but I personally find it more readable :)

File: src/haxelib/client/Main.hx (hunk 9 above)
Comment (ibilon) [inline src/haxelib/client/Main.hx]: Modified it and then returned it back, thought the spaces where from before, I'll return it to normal in the squash.

File: src/haxelib/SemVer.hx (hunk 6 above)
Comment (ibilon) [inline src/haxelib/SemVer.hx]: Would it be less weird if I added the other five?

Comment (nadako) [inline src/haxelib/SemVer.hx]: It seems reasonable to have comparison operators for SemVer, though I usually follow "refactor is used more than twice" rule :) I don't have a strong opinion here...

**Comment by ibilon** [comment]
Successfully squashed without changing pull request this time :)

Although I see that I brought run.n

**Comment by nadako** [comment]
Nice!

> Although I see that I brought run.n

That's how it should be now, since we want haxelib to be easily installable directly from git too :) Good thing it doesn't change between recompiles if source hasn't changed.

**Comment by nadako** [comment]
Merged by hand since it had conflicts with my earlier changes, sorry!

**Comment by ibilon** [comment]
Np ;)

**Comment by nadako** [comment]
> (update installs the latest version even if the lib isn't installed)

Turns out you broke this with `case []: throw "No such project " + project + " installed"` in the `projectNameToDir` function. I changed that to simply `return project`.
